packages/react-compose/src/next/README.md

```diff
@@ -0,0 +1,226 @@
+
+With these building blocks, you can compose or recompose the component in numerous ways.
+
+### Simple example
+
+A render function:
+
+```jsx
+const renderButton = state => {
+  return <button {...state} />;
+};
+```
+
+A hook which can manipulate the state (defining accessibility and behaviors):
+
+```jsx
+const useButton = state => {
+  // If the button is rendered as something besides a button or anchor, make it focusable.
+  if (state.as !== 'button' && state.as !== 'a') {
+    state.tabIndex = 0;
+  }
+};
+```
+
+A factory function sets up the state and render function:
+
+```jsx
+const createButton = (userProps, ref, defaultProps) => {
+  const state = _.merge({}, defaultProps, userProps);
+
+  // Apply button behaviors.
+  useButton(state);
+
+  return { state, render };
+};
+```
+
+A button can now be easily scaffolded, along with your choice of styling systems:
+
+```jsx
+import * as styles from './Button.scss';
+
+const useStyles = makeStyles(styles);
+
+const Button = React.forwardRef((props, ref) => {
+  const { state, render } = createButton(props, ref);
+
+  // Inject classNames as needed.
+  useStyles(state);
+
+  return render(state);
+});
+```
+
+### Details
+
+#### Creating mutable state with `mergeProps`
+
+In the previous example, `_.merge` was used to deep clone the props into a state object. Creating a single clone and using that to construct state simplifies hook development and usage; rather than trying to re-clone objects unnecessarily on every small mutation, hooks can assume operating against a draft state. This creates more self contained hooks, which can ensure they are used correctly, avoiding accidents like stomping on existing event handlers by blind object assigning the results.
+
+However, deep merge overlooks many edge cases for component props:
+
+- Deep merging classnames should append them, not replace
+- Deep merging JSX, ref objects, or arrays should replace, not recurse/clone
+- Deep merging an object on a string should replace
+
+...which introduces the first utility: `mergeProps`. Merge props works like a deep merge, but takes care of classnames, JSX, arrays, and object edge cases.
+
+#### Supporting the `as` prop with `getSlots`
+
+Fluent UI components take a common `as` prop. This allows the root element to be rendered with something other than the default.
+
+To support the `as` prop, the render function might look like this:
+
+```jsx
+const renderButton = state => {
+  const root = state.as;
+
+  return <root {...state} />;
+};
+```
+
+Additionally, you will need to filter out native properties which apply to the root; otherwise you will end up mixing any unexpected props into the element. To do this, we have a `getNativeElementProps` helper, which can be used for this purpose:
+
+```jsx
+const renderButton = state => {
+  const root = state.as;
+  const rootProps = getNativeElementProps(root, state);
```

Comment:
```jsx
const Link = React.forwardRef((props, ref) => {
  return <a {...props} ref={ref} />;
});

//

const root = state.as; // { as: Link }
const rootProps = getNativeElementProps(root, state);
```

What will be `rootProps` in this case?

Reply:
Assuming that as is a keyof html element names. I might have the typings wrong, but that's the intent.

Reply:
`as={ Link }` doesn't really make sense. `as={ AnythingWithTotallyDifferentProps }` would not make sense in typings land. If `Link` took in `foo` and `bar` as props, you would get errors.

What does make sense is to keep `as` restricted to intrinsic elements. We can always extend props to take in AllHTMLAttributes and that will always work.